File: 04_AJIN/backend/schemas/draft.py
"""초안 생성 관련 Pydantic 스키마.

Day 8 Phase 1 — 5 신규 엔드포인트용 스키마 추가:
- DocTypeListResponse, DocTypeMeta
- DraftStreamRequest
- CCRecRequest, CCRecResponse, CCGroup
- QualityRequest, QualityResponse, QualityScores, QualityIssue
- DiffRequest, DiffResponse, DiffStats

기존 4개 모델 (DraftGenerateRequest 등)은 backwards compatibility 위해 보존.
"""

from typing import Any, Literal

from pydantic import BaseModel, Field


# ═══════════════════════════════════════════════════════════
# 기존 모델 (보존)
# ═══════════════════════════════════════════════════════════


class DraftGenerateRequest(BaseModel):
    user_input: str
    doc_type: str | None = None
    tone: str = "공식적"
    include_ref: bool = True
    model: str | None = None
    language: str = "ko"
    recipient: str = "사내"
    context: str = "external"  # "internal" | "external"


class DraftResponse(BaseModel):
    session_id: str = ""
    doc_type: str = ""
    content: str = ""


class DraftReviseRequest(BaseModel):
    session_id: str
    instruction: str


class DraftExportRequest(BaseModel):
    content: str
    doc_type: str = "email_oem"
    format: str = "docx"  # "docx" | "pdf" | "hwpx" | "txt" | "odt" | "xlsx" | "csv"


# ═══════════════════════════════════════════════════════════
# Day 8 Phase 1 — 신규 5 엔드포인트 스키마
# ═══════════════════════════════════════════════════════════

# ── 1. GET /draft/doc-types ───────────────────────────────────


class DocTypeMeta(BaseModel):
    """13 문서 유형 메타 (canonical Draft.jsx 매핑)."""

    id: str
    category: Literal["internal", "external"]
    name_ko: str
    name_en: str = ""
    required_fields: list[str] = Field(default_factory=list)


class DocTypeListResponse(BaseModel):
    items: list[DocTypeMeta]
    internal_count: int = 0
    external_count: int = 0


# ── 2. POST /draft/stream (SSE) ───────────────────────────────


class DraftStreamRequest(BaseModel):
    """SSE 스트리밍 초안 생성 요청."""

    doc_type: str
    tone: str = "공식적"
    meta: dict[str, Any] = Field(default_factory=dict)  # title, recipient, content_request 등
    language: Literal["ko", "en"] = "ko"
    context: Literal["internal", "external"] = "internal"
    model: str | None = None


# ── 3. POST /draft/cc/recommend ───────────────────────────────


class CCRecRequest(BaseModel):
    """CC 자동 추천 요청 — features/draft/cc_recommender.recommend_cc 매핑."""

    doc_type: str
    sender_department: str = ""
    sender_division: str = ""
    recipient: str = ""  # 향후 확장용 (현재 cc_recommender 미사용)


class CCGroup(BaseModel):
    """CC 그룹 (필수/권장/선택)."""

    tier: Literal["required", "recommended", "optional"]
    label_ko: str
    label_en: str
    departments: list[str] = Field(default_factory=list)


class CCRecResponse(BaseModel):
    """3-tier CC 추천 결과."""

    groups: list[CCGroup]
    doc_type: str
    sender_department: str = ""


# ── 4. POST /draft/quality/score ──────────────────────────────


class QualityRequest(BaseModel):
    text: str
    doc_type: str
    reference_template: str = ""


class QualityScoresDetail(BaseModel):
    """5기준 점수 (각 max 다름)."""

    structure: float = 0.0  # 0~25
    structure_max: int = 25
    length: float = 0.0  # 0~20
    length_max: int = 20
    terminology: float = 0.0  # 0~25
    terminology_max: int = 25
    completeness: float = 0.0  # 0~15
    completeness_max: int = 15
    tone: float = 0.0  # 0~15
    tone_max: int = 15


class QualityResponse(BaseModel):
    """문서 품질 평가 응답 — features/draft/doc_quality_scorer.evaluate_document 매핑."""

    total_score: float = 0.0  # 0~100
    grade: str = "C"  # A / B+ / B / C / D / F
    scores: QualityScoresDetail = Field(default_factory=QualityScoresDetail)
    improvements: list[str] = Field(default_factory=list)
    details: dict[str, Any] = Field(default_factory=dict)


# ── 5. POST /draft/diff ───────────────────────────────────────


class DiffRequest(BaseModel):
    old: str
    new: str
    context_lines: int = 3


class DiffStats(BaseModel):
    added: int = 0
    removed: int = 0
    unchanged: int = 0
    similarity: float = 0.0  # 0.0 ~ 1.0


class DiffLine(BaseModel):
    """개별 diff 라인 (Frontend 렌더링용 — lg-diff-line.{add/del/mod/ctx})."""

    type: Literal["add", "del", "mod", "ctx", "header"]
    text: str


class DiffResponse(BaseModel):
    lines: list[DiffLine]
    stats: DiffStats
    diff_html: str = ""  # legacy HTML (선택)


# ═══════════════════════════════════════════════════════════
# Plan v1.0 — Module B 진단 / 모델 셀렉터 / SSE v2
# ═══════════════════════════════════════════════════════════


class DiagnoseCheck(BaseModel):
    """단일 의존성 점검 결과."""

    ok: bool
    detail: str = ""
    meta: dict[str, Any] = Field(default_factory=dict)


class DiagnoseResponse(BaseModel):
    """Module B 진단 — UI 헬스 배너용. 5개 항목.

    - ollama: 로컬 LLM 서버 가동 여부 + 설치 모델
    - gemini: Gemini API 키 존재 여부 (Feature B 에서는 차단되지만 진단은 노출)
    - pipeline: ENABLE_FEATURE_B + DraftPipeline 부팅 여부
    - templates: data/templates 의 .j2 템플릿 수
    - prompts: features/draft/prompts 의 시스템 프롬프트 수
    """

    ollama: DiagnoseCheck
    gemini: DiagnoseCheck
    pipeline: DiagnoseCheck
    templates: DiagnoseCheck
    prompts: DiagnoseCheck
    summary_ok: bool = False


class LLMOption(BaseModel):
    """모델 셀렉터 한 항목."""

    provider: Literal["ollama", "gemini"]
    id: str
    label: str
    available: bool = True
    blocked: bool = False  # Feature B 에서 보안상 차단됨
    blocked_reason: str = ""
    # v3.3 Feature C — exaone 패밀리 추가 (한국어 특화)
    family: Literal["qwen", "gemma", "gemini", "exaone", "other"] = "other"


class LLMOptionsResponse(BaseModel):
    options: list[LLMOption]
    default_provider: Literal["ollama", "gemini"] | None = None
    default_id: str | None = None
    feature: str = "draft"


class DraftStreamV2Request(BaseModel):
    """SSE v2 — Jinja2 템플릿 + RAG + LLMRouter 통합 흐름.

    기존 DraftStreamRequest 와 호환되며, provider/model 명시 + render_template 토글 추가.

    v3.6: reference_template_text — 사용자가 업로드한 참조 양식 (DOCX/PDF/HWP/TXT 추출 후
    POST /draft/upload-reference 의 응답 텍스트). 비어있지 않으면 LLM 프롬프트에 강력한
    "이 양식 그대로 따르세요" 지시문으로 prepend 됨.
    """

    doc_type: str
    tone: str = "공식적"
    meta: dict[str, Any] = Field(default_factory=dict)
    language: Literal["ko", "en"] = "ko"
    context: Literal["internal", "external"] = "internal"
    user_request: str = ""
    provider: Literal["ollama", "gemini"] | None = None
    model: str | None = None
    render_template: bool = True  # False 면 자유형 LLM 출력만 (호환 모드)
    reference_template_text: str = ""  # v3.6: 사용자 업로드 양식 (텍스트 추출본)
    reference_template_name: str = ""  # 원본 파일명 (UI 표시용)


class UploadReferenceResponse(BaseModel):
    """POST /draft/upload-reference 응답."""

    ok: bool
    filename: str
    extracted_chars: int
    truncated: bool  # UPLOAD_MAX_TEXT_CHARS 초과 시 잘림
    text: str  # 추출 텍스트 (LLM 프롬프트에 그대로 주입)
    detected_format: str  # docx / pdf / hwp / txt / md / unsupported
    warning: str = ""  # 부분 추출 등 경고

File: 04_AJIN/frontend/src/types/draft.ts
// Day 8 Phase 2 — Draft (Module B) TS 타입.
// backend/schemas/draft.py 의 Pydantic 모델과 1:1 정합.

export type DocCategory = 'internal' | 'external';
export type DraftPageTab = 'internal' | 'external' | 'history';
export type DraftToneId =
  | 'formal_internal'
  | 'formal_external'
  | 'standard'
  | 'friendly'
  | 'concise'
  | '공식적'
  | '친근함'
  | '중립';
export type CCTier = 'required' | 'recommended' | 'optional';

// ── 1. Doc Types (GET /draft/doc-types) ─────────────────────

export interface DocTypeMeta {
  id: string;
  category: DocCategory;
  name_ko: string;
  name_en: string;
  required_fields: string[];
}

export interface DocTypeListResponse {
  items: DocTypeMeta[];
  internal_count: number;
  external_count: number;
}

// ── 2. Stream Request (POST /draft/stream) ──────────────────

export interface DraftStreamRequest {
  doc_type: string;
  tone: string;
  meta: Record<string, unknown>;
  language?: 'ko' | 'en';
  context?: DocCategory;
  model?: string | null;
}

// ── 3. CC Recommend (POST /draft/cc/recommend) ──────────────

export interface CCRecRequest {
  doc_type: string;
  sender_department?: string;
  sender_division?: string;
  recipient?: string;
}

export interface CCGroup {
  tier: CCTier;
  label_ko: string;
  label_en: string;
  departments: string[];
}

export interface CCRecResponse {
  groups: CCGroup[];
  doc_type: string;
  sender_department: string;
}

// ── 4. Quality Score (POST /draft/quality/score) ────────────

export interface QualityRequest {
  text: string;
  doc_type: string;
  reference_template?: string;
}

export interface QualityScoresDetail {
  structure: number;
  structure_max: number;
  length: number;
  length_max: number;
  terminology: number;
  terminology_max: number;
  completeness: number;
  completeness_max: number;
  tone: number;
  tone_max: number;
}

export interface QualityResponse {
  total_score: number;
  grade: 'A' | 'B+' | 'B' | 'C' | 'D' | 'F';
  scores: QualityScoresDetail;
  improvements: string[];
  details: Record<string, unknown>;
}

// ── 5. Diff (POST /draft/diff) ──────────────────────────────

export interface DiffRequest {
  old: string;
  new: string;
  context_lines?: number;
}

export interface DiffStats {
  added: number;
  removed: number;
  unchanged: number;
  similarity: number;
}

export type DiffLineType = 'add' | 'del' | 'mod' | 'ctx' | 'header';

export interface DiffLine {
  type: DiffLineType;
  text: string;
}

export interface DiffResponse {
  lines: DiffLine[];
  stats: DiffStats;
  diff_html?: string;
}

// ── 6. Export (기존 POST /draft/export) ─────────────────────

export type ExportFormat = 'docx' | 'odt' | 'pdf' | 'xlsx' | 'csv' | 'txt' | 'hwpx';

export interface DraftExportRequest {
  content: string;
  doc_type?: string;
  format: ExportFormat;
}

// ── 7. Frontend 도메인 모델 (Firestore documents/{uid}/items/{id}) ──

export interface DraftDocumentMeta {
  title?: string;
  recipient?: string;
  cc?: string[];
  custom_fields?: Record<string, string>;
}

export interface DraftQualityRecord {
  total_score: number;
  grade: string;
  scores?: QualityScoresDetail;
}

export interface DraftVersionEntry {
  content: string;
  _at: number; // ms epoch
  source: 'llm' | 'user';
}

export interface DraftStorageUrls {
  docx?: string;
  odt?: string;
  pdf?: string;
  xlsx?: string;
  csv?: string;
  txt?: string;
}

export interface DraftDocument {
  id: string;
  user_uid: string;
  doc_type: string;
  tone: string;
  context: DocCategory;
  meta: DraftDocumentMeta;
  content: string;
  quality?: DraftQualityRecord;
  cc_recommended?: CCGroup[];
  versions: DraftVersionEntry[];
  storage_urls?: DraftStorageUrls;
  status: 'draft' | 'completed' | 'archived';
  created_at: number;
  updated_at: number;
}

// ═══════════════════════════════════════════════════════════
// Plan v1.0 — 진단 / 모델 옵션 / Stream v2
// ═══════════════════════════════════════════════════════════

export interface DiagnoseCheck {
  ok: boolean;
  detail: string;
  meta?: Record<string, unknown>;
}

export interface DiagnoseResponse {
  ollama: DiagnoseCheck;
  gemini: DiagnoseCheck;
  pipeline: DiagnoseCheck;
  templates: DiagnoseCheck;
  prompts: DiagnoseCheck;
  summary_ok: boolean;
}

export type LLMProvider = 'ollama' | 'gemini';
export type LLMFamily = 'qwen' | 'gemma' | 'gemini' | 'other';

export interface LLMOption {
  provider: LLMProvider;
  id: string;
  label: string;
  available: boolean;
  blocked: boolean;
  blocked_reason: string;
  family: LLMFamily;
}

export interface LLMOptionsResponse {
  options: LLMOption[];
  default_provider: LLMProvider | null;
  default_id: string | null;
  feature: string;
}

export interface DraftStreamV2Request {
  doc_type: string;
  tone: string;
  meta: Record<string, unknown>;
  language?: 'ko' | 'en';
  context?: DocCategory;
  user_request?: string;
  provider?: LLMProvider | null;
  model?: string | null;
  render_template?: boolean;
  /** v3.6 — 사용자 업로드 양식 텍스트 (POST /draft/upload-reference 응답 text) */
  reference_template_text?: string;
  /** v3.6 — 원본 파일명 (UI 표시용, 서버에서 프롬프트 파일명 라벨로 사용) */
  reference_template_name?: string;
}

/** POST /draft/upload-reference 응답 (v3.6 — 사용자 양식 업로드 후 텍스트 추출). */
export interface UploadReferenceResponse {
  ok: boolean;
  filename: string;
  extracted_chars: number;
  truncated: boolean;
  text: string;
  detected_format: 'docx' | 'pdf' | 'hwp' | 'hwpx' | 'txt' | 'md' | 'unsupported';
  warning: string;
}

export type StageStatus = 'running' | 'ok' | 'warn' | 'error';

export interface StageEvent {
  name: string;
  status: StageStatus;
  meta?: Record<string, unknown>;
}
